libevmasm/Assembly.cpp

```diff
@@ -971,6 +972,84 @@ void appendBigEndianUint16(bytes& _dest, ValueT _value)
 	assertThrow(_value <= 0xFFFF, AssemblyException, "");
 	appendBigEndian(_dest, 2, static_cast<size_t>(_value));
 }
+
+uint16_t calculateMaxStackHeight(AssemblyItems const& _items, uint16_t _args)
+{
+	static auto constexpr UNVISITED = std::numeric_limits<size_t>::max();
+
+	solAssert(!_items.empty());
+	uint16_t maxStackHeight = _args;
+	std::stack<size_t> worklist;
+	std::vector<size_t> stackHeights(_items.size(), UNVISITED);
+
+	// Init first item stack height to number of inputs to the code section
+	// stackHeights stores stack height for an item before the item execution
+	stackHeights[0] = _args;
+	// Push first item index to the worklist
+	worklist.push(0u);
+	while (!worklist.empty())
+	{
+		size_t idx = worklist.top();
+		worklist.pop();
+		AssemblyItem const& item = _items[idx];
+		size_t stackHeightChange = item.deposit();
+		size_t stackHeight = stackHeights[idx];
+		solAssert(stackHeight != UNVISITED);
+
+		std::vector<size_t> successors;
+
+		// Add next instruction to successors for non-changing control flow instructions
+		if (!(item.hasInstruction() && SemanticInformation::terminatesControlFlow(item.instruction())) &&
+			item.type() != RelativeJump &&
+			item.type() != RetF &&
+			item.type() != ReturnContract &&
+			item.type() != JumpF
+		)
+		{
+			solAssert(idx < _items.size() - 1, "No terminating instruction.");
+			successors.emplace_back(idx + 1);
+		}
+
+		// Add jumps destinations to successors
+		if (item.type() == RelativeJump || item.type() == ConditionalRelativeJump)
```

Comment:
Maybe we could assert that, if it's a backward jump, successor's stack height is not `UNVISITED`? Currently we just assume that `UNVISITED` means it cannot be a backward jump, but I think it would be useful to have a sanity check verifying it.

Reply:
Good point. Added